### ISSUE TITLE
reassembly of fragmented lwaftr packet:apps/lwaftr/fragmentv6_hardened.lua:158: Impossible case reached in v6 reassembly

### ISSUE THREAD

**Issue by chrgraf** [Igalia/snabbswitch#453]
Dear all,

I am running snabb lwaftr-snabbvmx-v1.0 standalone. While testing IPv6 reassembly, snabb cores with "apps/lwaftr/fragmentv6_hardened.lua:158: Impossible case reached in v6 reassembly"

Issue is reproducible in mine lab.
# How to reproduce

typically it is enough to send <20 fragmented  packets to lwaftr for a valid binding-entry. In mine case I am sending 10 times below valid fragmented packet:

```
06:46:41.800543 66:66:66:66:66:01 > 5c:45:27:15:a0:0d, ethertype IPv6 (0x86dd), length 494: 2a02:587:f710::40 > 2a02:587:f700::100: frag (0|432) truncated-ip - 796 bytes missing! 10.10.0.0.1024 > 10.0.1.100.17: UDP, length 1200
06:46:41.816526 66:66:66:66:66:01 > 5c:45:27:15:a0:0d, ethertype IPv6 (0x86dd), length 494: 2a02:587:f710::40 > 2a02:587:f700::100: frag (432|432)
06:46:41.828551 66:66:66:66:66:01 > 5c:45:27:15:a0:0d, ethertype IPv6 (0x86dd), length 426: 2a02:587:f710::40 > 2a02:587:f700::100: frag (864|364)
```

As long as snabb does not crash, it reassembles and decapsulates the lwaftr packet correctly:

```
06:46:41.828655 5c:45:27:15:a0:0d > 90:e2:ba:94:2a:bc, ethertype IPv4 (0x0800), length 1242: 10.10.0.0.1024 > 10.0.1.100.17: UDP, length 1200 
```

pcap is attached
# topology

ubuntu2-----router------snabb

with 
ubuntu2: IP 2a02:587:f710::40, 66:66:66:66:66:1
routers lwaftr mac: 5c:45:27:15:a0:0d
snabb tunnel-endpoint:2a02:587:f700::100 - see below config-file snabbvmx-lwaftr-xe1.conf

used pcap: 
[snabb-issue.pcap.gz](https://github.com/Igalia/snabb/files/489817/snabb-issue.pcap.gz)

binding-table:
[zipped.binding_table.txt.s.gz](https://github.com/Igalia/snabb/files/489827/zipped.binding_table.txt.s.gz)

```
lab@ubuntu1:~/standalone-ipv6-fragtest$ sudo usr/local/bin/snabb  lwaftr run --conf snabbvmx-lwaftr-xe1.conf --on-a-stick 0000:81:00.1
Warning: No NUMA memory affinity.
Pass --cpu to bind to a CPU and its NUMA node.
loading compiled binding table from ./binding_table.txt.s.o
compiled binding table ./binding_table.txt.s.o is up to date.
apps/lwaftr/fragmentv6_hardened.lua:158: Impossible case reached in v6 reassembly
stack traceback:
    core/main.lua:130: in function <core/main.lua:128>
    [C]: in function 'error'
    apps/lwaftr/fragmentv6_hardened.lua:158: in function 'attempt_reassembly'
    apps/lwaftr/fragmentv6_hardened.lua:286: in function 'cache_fragment'
    apps/lwaftr/ipv6_apps.lua:77: in function 'method'
    core/app.lua:87: in function 'with_restart'
    core/app.lua:335: in function 'thunk'
    core/histogram.lua:98: in function 'breathe'
    core/app.lua:273: in function 'main'
    program/lwaftr/run/run.lua:198: in function 'run'
    program/lwaftr/lwaftr.lua:17: in function 'run'
    core/main.lua:49: in function <core/main.lua:36>
    [C]: in function 'xpcall'
    core/main.lua:172: in main chunk
    [C]: at 0x00452340
    [C]: in function 'pcall'
    core/startup.lua:3: in main chunk
    [C]: in function 'require'
    [string "require "core.startup""]:1: in main chunk
```
# snabb-config

```
lab@ubuntu1:~/standalone-ipv6-fragtest$ cat snabbvmx-lwaftr-xe1.conf
hairpinning = false,
vlan_tagging = false,
binding_table = binding_table.txt.s,
aftr_ipv6_ip = fc00:168:5::2,
aftr_ipv4_ip = 192.168.5.2,
aftr_mac_inet_side = 02:cf:69:15:81:01,
inet_mac = 5c:45:27:15:a0:08,
aftr_mac_b4_side = 02:cf:69:15:81:01,
next_hop6_mac = 5c:45:27:15:a0:08,
ipv4_mtu = 9000,
ipv6_mtu = 9000,
```

as the binding-table is quite large(2mio entries), just the related stuff here. full binding-table attached above.

```
lab@ubuntu1:~/standalone-ipv6-fragtest$ less binding_table.txt.s
psid_map {
  10.10.0.0 {psid_length=6, shift=10}
  10.10.0.1 {psid_length=6, shift=10}
...
br_addresses {
  2a02:587:f700::100,
}
softwires {
  { ipv4=10.10.0.0, psid=1, b4=2a02:587:f710::40, aftr=0 }
  { ipv4=10.10.0.0, psid=2, b4=2a02:587:f710::41, aftr=0 }
  { ipv4=10.10.0.0, psid=3, b4=2a02:587:f710::42, aftr=0 }
  { ipv4=10.10.0.0, psid=4, b4=2a02:587:f710::43, aftr=0 }
...
```
# generating the packets

```
#!/bin/sh
e=/home/lab/cg-ubuntu2/packetget.py
v6="-v6 True -v6aftr_mac 5c:45:27:15:a0:0d -v6hgw_ip 2a02:587:f710::40 -v6aftr_ip 2a02:587:f700::100 -v6_len 1200 -v6frags 500"
v4="-v4 True -v4aftr_mac 5c:45:27:15:a0:0d -v4_len 2000 -v4frags 600"
common="-dev p4p1 -tf 1 -pr False -cip 10.10.0.0 -pip 32"

# binding-table
#$e $common $v4 $v6 -jcli binding_table.txt

#launch ipv4
#$e $common $v4
for i in `seq 1 5`;
do
$e $common $v6
done
```

**Comment by kbara** [comment]
To my great surprise, this is actually fixed by the work I did to fix the upstream Snabb definition of ntohl/htonl, before I looked into this bug. https://github.com/snabbco/snabb/pull/1024 has the gory details.

Specifically, get_frag_id is implemented this way:

```
local function get_frag_id(frag)
   local o_id = ehs + ipv6_fixed_header_size + o_ipv6_frag_id
   return ntohl(rd32(frag.data + o_id))
end
```

ntohl was returning a sign-extended 64 bit number: `ffffffff88f416c6`, while the reassembly buffer's cached version is an unsigned 32 bit integer as it should be, and ffffffff88f416c6 is not equal to 88f416c6 when more than 32 bits are being compared.

We'll patch it locally before upstream fixes it.

**Comment by kbara** [comment]
To be entirely clear: the problem isn't the number of packets to reassemble, the problem is a broken implementation of ntohl causing problems for packets which have their high bit set after ntohl.

**Comment by kbara** [comment]
Also: thank you for the excellent bug report, and clearly presenting the information to reproduce it. May I add the pcap to our test suite?

**Comment by chrgraf** [comment]
Hi Katerina,

Many thanks for this quick fix and your support. Much appreciated

Thanks

Christian

---

Christian Graf
Senior Consulting Engineer
Telephone +49 5651/336174
Mobile +49 175/2929 167
cgraf@juniper.net
www.juniper.net

Juniper Networks GmbH
Sitz der Gesellschaft / Registered office: Oskar-Schlemmer-Str. 15, 80807 München
Registergericht / Commercial registry: München, HRB 205571
Geschäftsführer / Directors: Johannes Albers, Stephen Byrne, Terrance Floyd Spidell

---

This email contains material that is confidential. The content of this email is for the sole use of the intended recipient(s).
Any review or distribution by persons other than the intended recipient(s) without the express permission of Juniper Networks is strictly prohibited.
If you are not the intended recipient, please contact the sender and delete/destroy all copies of this email and any related attachments.
Juniper does not guarantee the accuracy or completeness of third party materials or information

**Comment by kbara** [comment]
Welcome, Christian. Any version of the lwaftr >= git revision  e72231b8057d6cf2ff72b0de79b21e043429826c should now be ok with regards to this bug. :-)

**Comment by chrgraf** [comment]
Hi Katerina,

I verified your fix in in version 1.01 and it works like a charm. thanks

**Comment by kbara** [comment]
Glad to hear it. :)
